File: spec/dsn/dsn.go
// Package dsn parses loza:// connection URIs into resolved HTTP/HTTPS/WebSocket endpoints.
//
// The loza:// URI is the standard connection string for Loza Collector.
// It resolves to HTTP/HTTPS/OTLP/gRPC/WebSocket endpoints — it is NOT a wire protocol.
//
// Format:
//
//	loza://[username:password@][host][:port]/[collector]?env=<env>&service=<service>&tls=<true|false>&transport=<http|otlp|grpc>
//
// A private credential is username:password. A public bearer capability uses
// lx_pub_...: with an explicitly empty password.
// Userinfo credentials are percent-decoded and exposed in Username/Password.
// They are never included in resolved endpoint URLs.
//
// Examples:
//
//	loza://localhost:9308/demo?env=dev&tls=false
//	loza://key-id:s%40cret@collector.example.com/my-app?env=prod
//	loza://lx_pub_...:@collector.example.com/my-app?env=prod
//	loza://collector.example.com/my-app?env=prod&tls=true
//	loza://loza.internal:4318/backend?env=staging&service=auth&transport=otlp
package dsn

import (
	"fmt"
	"net/url"
	"strconv"
	"strings"
	"unicode"
)

// LozaDSN holds the parsed and resolved values from a loza:// connection URI.
type LozaDSN struct {
	Scheme        string // always "loza"
	Username      string // percent-decoded private key ID or public bearer capability
	Password      string // percent-decoded private key secret; empty for public capabilities
	Host          string // hostname (no port)
	Port          int    // resolved port number
	CollectorName string // canonical collector slug from the required path
	Project       string // deprecated compatibility alias for CollectorName
	Env           string // environment name (default: "default")
	Service       string // optional service name
	TLS           bool   // whether to use HTTPS
	Transport     string // "http", "otlp", or "grpc" (default: "http")
	BaseURL       string // resolved http(s)://host:port; never includes credentials
	EventsURL     string // base + /collectors/{collector}/events
	BatchURL      string // base + /collectors/{collector}/events/batch
	OTLPURL       string // base + /collectors/{collector}/otlp/logs
	TailWSURL     string // ws(s)://host:port/collectors/{collector}/tail
	LQLURL        string // base + /collectors/{collector}/lql/query
	LQLQueryURL   string // compatibility alias for LQLURL

}
// String returns a credential-free representation suitable for logs.
func (d LozaDSN) String() string {
	return d.BaseURL
}

// GoString returns a credential-free representation for %#v formatting.
func (d LozaDSN) GoString() string {
	return fmt.Sprintf("dsn.LozaDSN{BaseURL:%q}", d.BaseURL)
}

// Parse parses a raw loza:// connection URI into a LozaDSN.
//
// Validation rules:
//   - Scheme must be loza://
//   - Host is required (loza:// or loza:///project are rejected)
//   - Collector path is required (loza://host is rejected)
//   - Private userinfo must contain non-empty username/password
//   - Public userinfo is lx_pub_...: with an explicitly empty password
//   - Username cannot contain a colon or whitespace after decoding
//   - tls must be "true", "false", or "auto"
//   - transport must be "http", "otlp", or "grpc"
//   - Port must be 1-65535 if specified
//
// TLS defaults:
//   - localhost/127.0.0.1/::1 -> false
//   - everything else -> true
//
// Port defaults:
//   - tls=true -> 443
//   - tls=false -> 80
//   - localhost without explicit port -> 9308
func Parse(raw string) (*LozaDSN, error) {
	if raw == "" {
		return nil, fmt.Errorf("invalid Loza DSN: empty string")
	}

	if !strings.HasPrefix(raw, "loza://") {
		return nil, fmt.Errorf("invalid Loza DSN: scheme must be loza://")
	}

	// Parse as URL. The loza:// scheme is valid for url.Parse.
	u, err := url.Parse(raw)
	if err != nil {
		// Do not wrap URL parser details: malformed input may contain secrets.
		return nil, fmt.Errorf("invalid Loza DSN: malformed URL")
	}

	username := ""
	password := ""
	if u.User != nil {
		var hasPassword bool
		username = u.User.Username()
		password, hasPassword = u.User.Password()
		if !hasPassword || username == "" || (password == "" && !IsPublicCredentialUsername(username)) {
			return nil, fmt.Errorf("invalid Loza DSN: credentials require username:password or lx_pub_...:")
		}
		if strings.Contains(username, ":") || hasWhitespace(username) {
			return nil, fmt.Errorf("invalid Loza DSN: username contains an invalid character")
		}
		if err := validateRawPassword(raw); err != nil {
			return nil, err
		}
	}

	host := u.Hostname()
	if host == "" {
		return nil, fmt.Errorf("invalid Loza DSN: host is required")
	}

	portStr := u.Port()

	// The required path is the canonical collector identity. Project remains
	// available as a compatibility alias for existing SDK consumers.
	collectorName := strings.TrimPrefix(u.Path, "/")
	if collectorName == "" {
		return nil, fmt.Errorf("invalid Loza DSN: collector path is required, e.g. loza://host/my-collector")
	}

	q := u.Query()

	// ── TLS default ──────────────────────────────────────────────────────────
	tls := true
	if isLocalhost(host) {
		tls = false
	}
	if v := q.Get("tls"); v != "" {
		switch v {
		case "true":
			tls = true
		case "false":
			tls = false
		case "auto":
			// keep the computed default
		default:
			return nil, fmt.Errorf("invalid Loza DSN: tls must be true, false, or auto, got %q", v)
		}
	}

	// ── Port default ─────────────────────────────────────────────────────────
	port := 443
	if !tls {
		port = 80
	}
	if isLocalhost(host) && portStr == "" {
		port = 9308
	}
	if portStr != "" {
		p, err := strconv.Atoi(portStr)
		if err != nil || p < 1 || p > 65535 {
			return nil, fmt.Errorf("invalid Loza DSN: invalid port %q", portStr)
		}
		port = p
	}

	// ── Transport ────────────────────────────────────────────────────────────
	transport := "http"
	if v := q.Get("transport"); v != "" {
		switch v {
		case "http", "otlp", "grpc":
			transport = v
		default:
			return nil, fmt.Errorf("invalid Loza DSN: transport must be http, otlp, or grpc, got %q", v)
		}
	}

	// ── Env ──────────────────────────────────────────────────────────────────
	env := q.Get("env")
	if env == "" {
		env = "default"
	}

	service := q.Get("service")

	// ── Build resolved URLs ──────────────────────────────────────────────────
	scheme := "https"
	wsScheme := "wss"
	if !tls {
		scheme = "http"
		wsScheme = "ws"
	}

	// IPv6 addresses must be bracketed in URLs per RFC 2732/3986.
	hostPart := host
	if strings.Contains(host, ":") {
		hostPart = "[" + host + "]"
	}

	baseURL := fmt.Sprintf("%s://%s:%d", scheme, hostPart, port)

	collectorPath := url.PathEscape(collectorName)
	collectorBaseURL := baseURL + "/collectors/" + collectorPath
	collectorTailBaseURL := fmt.Sprintf("%s://%s:%d/collectors/%s", wsScheme, hostPart, port, collectorPath)

	return &LozaDSN{
		Scheme:        "loza",
		Username:      username,
		Password:      password,
		Host:          host,
		Port:          port,
		CollectorName: collectorName,
		Project:       collectorName,
		Env:           env,
		Service:       service,
		TLS:           tls,
		Transport:     transport,
		BaseURL:       baseURL,
		EventsURL:     collectorBaseURL + "/events",
		BatchURL:      collectorBaseURL + "/events/batch",
		OTLPURL:       collectorBaseURL + "/otlp/logs",
		TailWSURL:     collectorTailBaseURL + "/tail",
		LQLURL:        collectorBaseURL + "/lql/query",
		LQLQueryURL:   collectorBaseURL + "/lql/query",
	}, nil
}

func hasWhitespace(value string) bool {
	for _, r := range value {
		if unicode.IsSpace(r) {
			return true
		}
	}
	return false
}

// validateRawPassword enforces that URL-reserved password bytes are encoded.
// url.Parse has already validated percent escapes and decoded u.User.
func validateRawPassword(raw string) error {
	const prefix = "loza://"
	authorityEnd := len(raw)
	for i := len(prefix); i < len(raw); i++ {
		switch raw[i] {
		case '/', '?', '#':
			authorityEnd = i
			i = len(raw)
		}
	}

	authority := raw[len(prefix):authorityEnd]
	at := strings.LastIndexByte(authority, '@')
	if at < 0 {
		return fmt.Errorf("invalid Loza DSN: malformed credentials")
	}
	userinfo := authority[:at]
	colon := strings.IndexByte(userinfo, ':')
	if colon < 0 {
		return fmt.Errorf("invalid Loza DSN: credentials require non-empty username and password")
	}

	rawPassword := userinfo[colon+1:]
	for i := range rawPassword {
		if rawPassword[i] == '%' {
			if i+2 >= len(rawPassword) || !isHexDigit(rawPassword[i+1]) || !isHexDigit(rawPassword[i+2]) {
				return fmt.Errorf("invalid Loza DSN: malformed credentials")
			}
			continue
		}
		if isURLReserved(rawPassword[i]) {
			return fmt.Errorf("invalid Loza DSN: password contains an unencoded reserved character")
		}
	}
	return nil
}

func isURLReserved(value byte) bool {
	switch value {
	case ':', '/', '?', '#', '[', ']', '@', '!', '$', '&', '\'', '(', ')', '*', '+', ',', ';', '=':
		return true
	default:
		return false
	}
}

func isHexDigit(value byte) bool {
	return (value >= '0' && value <= '9') ||
		(value >= 'a' && value <= 'f') ||
		(value >= 'A' && value <= 'F')
}

// IsPublicCredentialUsername reports whether username is the public DSN bearer
// capability form. Its empty Basic password is intentional.
func IsPublicCredentialUsername(username string) bool {
	const prefix = "lx_pub_"
	return strings.HasPrefix(username, prefix) && len(username) > len(prefix)
}

// isLocalhost returns true for localhost, 127.0.0.1, or ::1.
func isLocalhost(host string) bool {
	return host == "localhost" || host == "127.0.0.1" || host == "::1"
}

File: sdks/py/src/loxa/core/dsn.py
"""Parse loxa:// connection URIs into resolved HTTP/HTTPS/WebSocket endpoints.

The loxa:// URI is the standard connection string for Loxa Collector.
It resolves to HTTP/HTTPS/OTLP/gRPC/WebSocket endpoints — it is NOT a wire protocol.

Format::

    loxa://[host][:port]/[project]?env=<env>&service=<service>&tls=<true|false|auto>&transport=<http|otlp|grpc>
"""

from __future__ import annotations

import os
from dataclasses import dataclass
from urllib.parse import urlparse, parse_qs


@dataclass(frozen=True, slots=True)
class LoxaDSN:
    """Parsed and resolved loxa:// connection URI."""

    scheme: str
    host: str
    port: int
    project: str
    env: str
    service: str
    tls: bool
    transport: str
    base_url: str
    events_url: str
    batch_url: str
    otlp_url: str
    tail_ws_url: str


_LOCALHOSTS = frozenset({"localhost", "127.0.0.1", "::1"})


def _is_localhost(host: str) -> bool:
    return host in _LOCALHOSTS


def parse(raw: str) -> LoxaDSN:
    """Parse a raw loxa:// connection URI into a LoxaDSN.

    Raises:
        ValueError: if the URI is invalid.
    """
    if not raw:
        raise ValueError("invalid Loxa DSN: empty string")

    if not raw.startswith("loxa://"):
        raise ValueError("invalid Loxa DSN: scheme must be loxa://")

    u = urlparse(raw)

    # Reject userinfo (API keys must not be in the URL).
    if u.username:
        raise ValueError(
            "invalid Loxa DSN: do not put API keys in the URL, use LOXA_API_KEY instead"
        )

    host = u.hostname or ""
    if not host:
        raise ValueError("invalid Loxa DSN: host is required")

    # Parse port — urlparse raises ValueError for non-numeric ports.
    try:
        port_raw = u.port
    except ValueError:
        raise ValueError("invalid Loxa DSN: invalid port")

    # Project is the path segment without leading slash.
    project = u.path.lstrip("/")
    if not project:
        raise ValueError(
            "invalid Loxa DSN: project path is required, e.g. loxa://host/my-project"
        )

    q = parse_qs(u.query)

    # --- TLS default --------------------------------------------------------
    tls: bool = not _is_localhost(host)
    tls_val = q.get("tls", [""])[0]
    if tls_val:
        if tls_val == "true":
            tls = True
        elif tls_val == "false":
            tls = False
        elif tls_val == "auto":
            pass  # keep computed default
        else:
            raise ValueError(
                f"invalid Loxa DSN: tls must be true, false, or auto, got {tls_val!r}"
            )

    # --- Port default -------------------------------------------------------
    if tls:
        port = 443
    else:
        port = 80

    if _is_localhost(host) and port_raw is None:
        port = 9308

    if port_raw is not None:
        if not (1 <= port_raw <= 65535):
            raise ValueError(f"invalid Loxa DSN: invalid port {port_raw!r}")
        port = port_raw

    # --- Transport ----------------------------------------------------------
    transport = "http"
    transport_val = q.get("transport", [""])[0]
    if transport_val:
        if transport_val in ("http", "otlp", "grpc"):
            transport = transport_val
        else:
            raise ValueError(
                f"invalid Loxa DSN: transport must be http, otlp, or grpc, got {transport_val!r}"
            )

    # --- Env ----------------------------------------------------------------
    env = q.get("env", ["default"])[0]
    if not env:
        env = "default"

    service = q.get("service", [""])[0]

    # --- Build resolved URLs ------------------------------------------------
    scheme = "https" if tls else "http"
    ws_scheme = "wss" if tls else "ws"

    # IPv6 addresses must be bracketed in URLs per RFC 2732/3986.
    host_part = f"[{host}]" if ":" in host else host

    base_url = f"{scheme}://{host_part}:{port}"

    return LoxaDSN(
        scheme="loxa",
        host=host,
        port=port,
        project=project,
        env=env,
        service=service,
        tls=tls,
        transport=transport,
        base_url=base_url,
        events_url=base_url + "/events",
        batch_url=base_url + "/events/batch",
        otlp_url=base_url + "/otlp/logs",
        tail_ws_url=f"{ws_scheme}://{host_part}:{port}/tail",
    )


def resolve_endpoint(path: str = "", default: str = "") -> str:
    """Resolve an endpoint from LOXA_COLLECTOR_URL env var with a fallback.

    Accepts both loxa:// DSN and plain http(s) URLs. If LOXA_COLLECTOR_URL
    is a loxa:// DSN, it is resolved to an HTTP(S) URL first.

    Args:
        path: Suffix appended to the base URL (e.g. "/events").
        default: Fallback value when LOXA_COLLECTOR_URL is not set.

    Returns:
        The resolved endpoint URL.
    """
    base = os.getenv("LOXA_COLLECTOR_URL", "").strip()
    if base:
        if base.startswith("loxa://"):
            try:
                base = parse(base).base_url
            except Exception:
                pass  # Use raw value; caller will handle the invalid URL
        return base.rstrip("/") + path
    return default
